lib/runner/variant-reassignment.js

```diff
@@ -282,18 +277,95 @@ export default class VariantReassignment {
   _deleteTransaction () {
   }
 
-  _deleteCtpProductToUpdate () {
+  /**
+   * Compare if two product objects are the same product.
+   * @private
+   */
+  _isProductsSame () {
+  }
+
+  _createNewProduct () {
   }
 
-  _thereAreConflictingSlugsInMatchingProducts () {
+  /**
+   * Create a backup of a product because we need to do product type change for this product
+   */
+  _backupProductForProductTypeChange () {
   }
 
-  _anonymizeProductSlugAndKey () {
+  _changeProductType () {
   }
 
-  _createNewProduct () {
+  /**
+   * Delete a backup that was created because of product type change of a product
+   */
+  _deleteBackupForProductTypeChange () {
+  }
+
+  /**
+   * Verify that there are no other products in the platform that has slugs of productDraft
+   * except ctpProductToUpdate.
+   * This method should cover the test variant-reassignment-example-7.spec.js
+   */
+  _ensureSlugUniqueness () {
   }
 
-  _addMissingVariants () {
+  async _removeVariantsFromCtpProductToUpdate (anonymizedProductDraft, ctpProductToUpdate) {
+    const skusToRemove = this.productService.getProductDraftSkus(anonymizedProductDraft)
+    await this.productService.removeVariantsFromProduct(ctpProductToUpdate, skusToRemove)
   }
+
+  async _createVariantsInCtpProductToUpdate (backupVariants, productDraft, ctpProductToUpdate) {
+    const actions = []
+    const skuToVariant = new Map()
+    const existingSkus = this.productService.getProductSkus(ctpProductToUpdate)
+    const variants = productDraft.variants || []
+    variants.concat(productDraft.masterVariant).forEach((v) => {
+      if (!existingSkus.includes(v.sku))
+        skuToVariant.set(v.sku, v)
+    })
+    // preserve existing attribute data
+    if (!_.isEmpty(this.retainExistingAttributes))
```

Comment:
Here we can create variants from backupVariants instead of productDraft so we automatically retain existing attributes. Later the product import will do the rest

Reply:
One problem could occur with your solution - if the productDraft has different product type, the backupVariants might have attributes which are not in the new product type anymore. If we create new variants with productDraft and merge only the defined attributes, we could avoid this problem.

Reply:
sounds ok to me .. but then you should preserve always attributes price and assets so the `retainExistingAttributes` will have always at least two values, right?

Reply:
It depends on what the user wants. If they want to preserve price and assets, they can set it. But if they have always price or assets with the product draft, so then they don't want to preserve.